File: backend/app/schemas/deployment.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class DeploymentPlanCreate(BaseModel):
    device_ids: list[int] = Field(min_length=1, max_length=5)


class DeviceReleaseOverrideSet(BaseModel):
    release_id: int = Field(ge=1)
    reason: str = Field(min_length=1, max_length=1000)
    expires_at: datetime


class DeviceReleaseOverrideRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    device_id: int
    function_id: int
    release_id: int
    reason: str
    expires_at: datetime
    active: bool
    created_by: int
    created_at: datetime
    updated_at: datetime


class DeploymentPlanItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    plan_id: int
    device_id: int
    function_id: int
    release_id: int
    variant_id: int
    config_snapshot: dict | None
    config_hash: str
    artifact_sha256: str
    preflight_json: dict | None
    status: str
    created_at: datetime


class DeploymentPlanRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    status: str
    snapshot_hash: str
    expires_at: datetime
    stale_reason: str | None
    created_by: int
    confirmed_by: int | None
    confirmed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    items: list[DeploymentPlanItemRead] = Field(default_factory=list)


class DeploymentExecutionItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    deployment_execution_id: int
    plan_item_id: int
    status: str
    attempt_count: int
    result_json: dict | None
    error_message: str | None
    started_at: datetime | None
    finished_at: datetime | None
    created_at: datetime
    updated_at: datetime


class DeploymentExecutionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    execution_id: str
    plan_id: int
    status: str
    started_at: datetime | None
    finished_at: datetime | None
    created_by: int
    created_at: datetime
    updated_at: datetime
    items: list[DeploymentExecutionItemRead] = Field(default_factory=list)

File: frontend/src/api/domains/deployment.ts
import { api } from "../core";

export interface DeploymentPlanItemRead {
  id: number;
  plan_id: number;
  device_id: number;
  function_id: number;
  release_id: number;
  variant_id: number;
  config_snapshot: Record<string, unknown> | null;
  config_hash: string;
  artifact_sha256: string;
  preflight_json: Record<string, unknown> | null;
  status: string;
  created_at: string;
}

export interface DeploymentPlanRead {
  id: number;
  project_id: number;
  status: string;
  snapshot_hash: string;
  expires_at: string;
  stale_reason: string | null;
  created_by: number;
  confirmed_by: number | null;
  confirmed_at: string | null;
  created_at: string;
  updated_at: string;
  items: DeploymentPlanItemRead[];
}

export interface DeploymentExecutionItemRead {
  id: number;
  deployment_execution_id: number;
  plan_item_id: number;
  status: string;
  attempt_count: number;
  result_json: Record<string, unknown> | null;
  error_message: string | null;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface DeploymentExecutionRead {
  id: number;
  execution_id: string;
  plan_id: number;
  status: string;
  started_at: string | null;
  finished_at: string | null;
  created_by: number;
  created_at: string;
  updated_at: string;
  items: DeploymentExecutionItemRead[];
}

export async function createDeploymentPlan(projectId: number, deviceIds: number[]): Promise<DeploymentPlanRead> {
  const response = await api.post<DeploymentPlanRead>(`/projects/${projectId}/deployment-plans`, {
    device_ids: deviceIds,
  });
  return response.data;
}

export async function getDeploymentPlan(planId: number): Promise<DeploymentPlanRead> {
  const response = await api.get<DeploymentPlanRead>(`/deployment-plans/${planId}`);
  return response.data;
}

export async function confirmDeploymentPlan(planId: number): Promise<DeploymentExecutionRead> {
  const response = await api.post<DeploymentExecutionRead>(`/deployment-plans/${planId}/confirm`);
  return response.data;
}

export async function getDeploymentExecution(executionId: string): Promise<DeploymentExecutionRead> {
  const response = await api.get<DeploymentExecutionRead>(`/deployment-executions/${executionId}`);
  return response.data;
}

export async function executeDeploymentExecution(executionId: string): Promise<DeploymentExecutionRead> {
  const response = await api.post<DeploymentExecutionRead>(`/deployment-executions/${executionId}/execute`);
  return response.data;
}
